### ISSUE TITLE
ServiceEntry merging behavior in 1.22.x

### ISSUE THREAD

**Issue by jarnfast** [istio/istio#52546]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

The way multiple ServiceEntries are merged have changed drastically in 1.22.x and I wonder if this really is the intended behavior.
The behavioral changes appear to originate in https://github.com/istio/istio/blob/1.22.3/pilot/pkg/model/sidecar.go#L1001

Previously we could have a ServiceEntry export to the entire mesh and then have additional ServiceEntries in relevant names extended the allowed ports.
Now the `metadata.namespace`, `spec.exportTo` and `metadata.labels` are used to determine if the ServiceEntries can be merged. (The bit with got me extra baffled)

The change notes for 1.22.0 does mention some changes around `SidecarScope` construction but no references to issues or reasoning behind the changes.

# Test Istio resources (setup.yaml)
```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
  name: test1
---
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
  name: test2
---
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: dummysite-p100-exported-to-all
  namespace: test1
spec:
  hosts:
  - dummysite.local
  location: MESH_EXTERNAL
  ports:
  - name: https-100
    number: 100
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: dummysite-p200-exported-to-self
  namespace: test2
spec:
  exportTo:
    - "."
  hosts:
  - dummysite.local
  location: MESH_EXTERNAL
  ports:
  - name: https-200
    number: 200
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  labels:
    app.kubernetes.io/part-of: secondary-service-in-same-namespace
  name: dummysite-p300-exported-to-self-and-with-label
  namespace: test2
spec:
  exportTo:
    - "."
  hosts:
  - dummysite.local
  location: MESH_EXTERNAL
  ports:
  - name: https-300
    number: 300
    protocol: HTTPS
  resolution: DNS
```

# How to reproduce

```sh
$ kubectl apply -f setup.yaml
$ kubectl run -n test1 test1app --image=busybox
$ kubectl run -n test2 test2app --image=busybox
```

# Results in Istio 1.21.5

Here we see that the 3 ServiceEntries are merged for the `test2app` pod (as we expected)

```sh
$ istioctl pc cluster test1app.test1 | grep -E "(^SERVICE|dummysite)"
SERVICE FQDN                                                           PORT      SUBSET     DIRECTION     TYPE             DESTINATION RULE
dummysite.local                                                        100       -          outbound      STRICT_DNS

$ istioctl pc cluster test2app.test2 | grep -E "(^SERVICE|dummysite)"
SERVICE FQDN                                                           PORT      SUBSET     DIRECTION     TYPE             DESTINATION RULE
dummysite.local                                                        100       -          outbound      STRICT_DNS
dummysite.local                                                        200       -          outbound      STRICT_DNS
dummysite.local                                                        300       -          outbound      STRICT_DNS
```

# Results in Istio 1.22.3

Here we see that only a single ServiceEntry is in effect for the `test2app` pod. The actual entry in effect is determined by the creation timestamp of the ServiceEntries - but never the one exported from the `test1` namespace.

```sh
$ istioctl pc cluster test1app.test1 | grep -E "(^SERVICE|dummysite)"
SERVICE FQDN                                                           PORT      SUBSET     DIRECTION     TYPE             DESTINATION RULE
dummysite.local                                                        100       -          outbound      STRICT_DNS

$ istioctl pc cluster test2app.test2 | grep -E "(^SERVICE|dummysite)"
SERVICE FQDN                                                           PORT      SUBSET     DIRECTION     TYPE             DESTINATION RULE
dummysite.local                                                        200       -          outbound      STRICT_DNS
```

_Note:_ If the ServiceEntries in `test2` namespace are removed completely then the ServiceEntry from `test1` namespace is exported correctly

### Version

```Text
$ istioctl version
client version: 1.22.3
control plane version: 1.22.3
data plane version: 1.22.3 (23 proxies)
$ kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.2+k3s1
```

**Comment by keithmattix** [comment]
The behavior of duplicate hosts is typically undefined in Istio; the previous ServiceEntry behavior was responsible for several bugs. Most of the context can be found in https://github.com/istio/istio/pull/50711 and linked issues within the description. Happy to chat more if you have other questions/concerns!

**Comment by jarnfast** [comment]
It may had been responsible for several bugs - but the new behavior is far from backwards compatible 😉 

In setups with multiple application teams we'd like to have some shared/common ServiceEntries and then let each team create additional ServiceEntries - in some cases with additional ports on predefined hosts. With the new behavior each team must also remember to include whatever ports the common ServiceEntry defined - and preferable only once per namespace?

Maybe we can get away with having each app create their ServiceEntries with workload selectors - assuming that those get prioritized over creation timestamp when there's duplicate hosts. 

I haven't found anything in the linked issues that tells me why `metadata.labels`  are used to determine if merging is possible?

In summary, I just want to be _really_ sure that the new behavior wasn't faulty before starting the required rework 😉

**Comment by keithmattix** [comment]
Ack, that's perfectly reasonable. Sorry for the trouble. I don't know if labels was purposefully used as part of merging criteria. Maybe @howardjohn or @hzxuzhonghu can speak more to it

**Comment by keithmattix** [comment]
Actually, looked like the behavior changed in #49595

**Comment by jarnfast** [comment]
> Actually, looked like the behavior changed in #49595

Yeah, I found that one as well - the guys are discussing the label selectors (ie. `.spec.workloadSelector.labels`) - yet the PR also included the labels (ie. `metadata.labels`) as part of the merge magic.

**Comment by keithmattix** [comment]
That feels like a bug... @hzxuzhonghu was this intentional?

**Comment by sivaramsajeev** [comment]
@keithmattix  is there a plan to modify the logic of merge? We faced issues with the merge logic where it skips the merge when labels are different. We do have multiple namespaces with more than a single app and same hosts defined in serviceentries. 

I see this is till unchanged in the new release https://github.com/istio/istio/blob/1.23.2/pilot/pkg/model/sidecar.go#L1014. So we are looking at fixing the charts now.

**Comment by keithmattix** [comment]
Looks like we reverted this fix: https://istio.io/latest/news/releases/1.22.x/announcing-1.22.5/#changes

**Comment by jarnfast** [comment]
> Looks like we reverted this fix: https://istio.io/latest/news/releases/1.22.x/announcing-1.22.5/#changes

@keithmattix thanks for the follow up - it's still unclear what the intended future behavior will be ;-) In the mean time we'll be reworking our use of ServiceEntries to avoid merging as much as possible.

**Comment by iampratapak** [comment]
We are also facing a similar issue. https://github.com/istio/istio/issues/53667

**Comment by hzxuzhonghu** [comment]
Yeah, i will take a deep look.
